src/main/java/graphql/validation/rules/OverlappingFieldsCanBeMerged.java

```diff
@@ -64,187 +68,163 @@ public void leaveSelectionSet(SelectionSet selectionSet) {
-        for (FieldPair fieldPair : alreadyChecked) {
-            if (fieldPair.field1 == field1 && fieldPair.field2 == field2) {
-                return true;
-            }
-            if (fieldPair.field1 == field2 && fieldPair.field2 == field1) {
-                return true;
+        for (Selection selection : selectionSet.getSelections()) {
+            if (selection instanceof Field) {
+                collectFieldsForField(fieldMap, parentType, (Field) selection);
+
+            } else if (selection instanceof InlineFragment) {
+                collectFieldsForInlineFragment(fieldMap, visitedFragmentSpreads, parentType, (InlineFragment) selection);
+
+            } else if (selection instanceof FragmentSpread) {
+                collectFieldsForFragmentSpread(fieldMap, visitedFragmentSpreads, (FragmentSpread) selection);
             }
         }
-        return false;
     }
 
-    @SuppressWarnings("ConstantConditions")
-    private Conflict findConflict(String responseName, FieldAndType fieldAndTypeA, FieldAndType fieldAndTypeB) {
-
-        Field fieldA = fieldAndTypeA.field;
-        Field fieldB = fieldAndTypeB.field;
-
-        if (isAlreadyChecked(fieldA, fieldB)) {
-            return null;
+    private void collectFieldsForFragmentSpread(Map<String, Set<FieldAndType>> fieldMap, Set<String> visitedFragmentSpreads, FragmentSpread fragmentSpread) {
+        FragmentDefinition fragment = getValidationContext().getFragment(fragmentSpread.getName());
+        if (fragment == null) {
+            return;
         }
-        alreadyChecked.add(new FieldPair(fieldA, fieldB));
-
-        String fieldNameA = fieldA.getName();
-        String fieldNameB = fieldB.getName();
-
-        GraphQLType typeA = fieldAndTypeA.graphQLType;
-        GraphQLType typeB = fieldAndTypeB.graphQLType;
-
-        Conflict conflict = checkListAndNonNullConflict(responseName,fieldAndTypeA,fieldAndTypeB);
-        if (conflict != null) {
-            return conflict;
+        if (visitedFragmentSpreads.contains(fragment.getName())) {
+            return;
         }
+        visitedFragmentSpreads.add(fragment.getName());
+        GraphQLType graphQLType = TypeFromAST.getTypeFromAST(getValidationContext().getSchema(),
+                fragment.getTypeCondition());
+        collectFields(fieldMap, fragment.getSelectionSet(), graphQLType, visitedFragmentSpreads);
+    }
 
-        typeA = unwrapAll(typeA);
-        typeB = unwrapAll(typeB);
+    private void collectFieldsForInlineFragment(Map<String, Set<FieldAndType>> fieldMap, Set<String> visitedFragmentSpreads, GraphQLType parentType, InlineFragment inlineFragment) {
+        GraphQLType graphQLType = inlineFragment.getTypeCondition() != null
+                ? TypeFromAST.getTypeFromAST(getValidationContext().getSchema(), inlineFragment.getTypeCondition())
+                : parentType;
+        collectFields(fieldMap, inlineFragment.getSelectionSet(), graphQLType, visitedFragmentSpreads);
```

Comment:
GraphQLType graphQLType = TypeFromAST.getTypeFromAST(getValidationContext().getSchema(),
                fragment.getTypeCondition());
 
 GraphQLType graphQLType = inlineFragment.getTypeCondition() != null
                ? TypeFromAST.getTypeFromAST(getValidationContext().getSchema(), inlineFragment.getTypeCondition())
                : parentType;
                
                
The above repeated code deserves a method

Reply:
done ...
just for the record: this is the same code as before: I didn't touch this part for this PR.

On a general note: this is very old code. In general there is a huge overlap in logic between FieldCollector, OverlappingField validation and the normalized field factory.